### ISSUE TITLE
Make the Map pickable

### ISSUE THREAD

**Issue by BastienGauthier** [python-visualization/folium#1813]
**Is your feature request related to a problem? Please describe.**
As an extension of https://github.com/python-visualization/branca/pull/99, I want to be able to cache a map for an application, to switch quickly between them.

**Describe the solution you'd like**
I proposed a first solution in the following PR : https://github.com/python-visualization/folium/pull/1812
The solution is partial and would greatly appreciate any help, as I feel a pickable Map could be helpful in many ways.

**Additional context**
Map before correction :
![image](https://github.com/python-visualization/folium/assets/108142642/5dd6ecbc-d154-4262-82fe-cc840c361a43)

Map after correction 
![image](https://github.com/python-visualization/folium/assets/108142642/615b1c89-f3d0-41d7-8957-a0e6d3ecc84e)

**Implementation**
See https://github.com/python-visualization/folium/pull/1812

**Comment by BastienGauthier** [comment]
Ok, seems to work now ! The solution proposed needs branca modification https://github.com/python-visualization/branca/pull/144 though.
I tested with some elements but I did not go though everything, even though a made the modification everywhere. If I export the html from the map, the files before pickling and after unpickling matches.

**Comment by Conengmo** [comment]
Interesting! I haven't looked into the changes very thoroughly, but I wanted to challenge first already:

Is it possible to do this with changes *just* in Branca's Element class? Not in every class? So no adding `template_str` and no adding `__set_state__` everywhere.
  - Since everything inherits from Element, changing getstate and setstate there should automatically work for every Folium class as well
  - Can you get the template string from an initialized `Template` object? That way there's no need to add `template_str` everywhere. You could extract the string from the `Template` object in getstate.

I haven't tried this myself so these are questions for you! Curious to hear what is possible.

**Comment by BastienGauthier** [comment]
Indeed, I just tried and I can still simplify the setstate I will update the PR soon ! 
Regarding the template string, I could not find any way to extract it. I will be happy to implement it if you find something on your side !

**Comment by Conengmo** [comment]
Too bad we can't extract the template string from the Template objects. I'd still like to avoid having to update every class though by specifying `template_str` everywhere.

Maybe as an alternative we can remove the `_template` object when pickling, then when unpickling re-add it from the class? When we call setstate, we know what class we have, and `_template` is available as a class attribute.

**Comment by BastienGauthier** [comment]
Not sure how to add it from the class actually : do you have an exemple ?

**Comment by Conengmo** [comment]
I looked into it a bit more, learned more about pickling and unpickling. Forget what I said in my previous comment. I learned that class attributes are not pickled, so we don't have to worry about `_template` as a class attribute. Only the `template` argument of `Element` is problematic, but I see you got that covered in https://github.com/python-visualization/branca/pull/144!

I also realised we can save ourselves some trouble by not adding `ENV` as an instance attribute in the first place. I don't see a good reason to that. We have it as a module attribute, why add it to class instances? It's a private attribute even, so we don't expect users to use it. Let's not do that in your open PR, but I propose we do a separate PR and don't add `ENV` to any classes.

@BastienGauthier is https://github.com/python-visualization/branca/pull/144 ready for review?
